#100 ゲームフロークラスの作成

t-nagatsuka-gr opened the pull request.

t-nagatsuka-gr committed:
#100 ゲームフロークラスの作成

a-kiyo-g-relations commented on GameFlow.cs:
```diff
@@ -0,0 +1,114 @@
+public class GameFlow
+{
+    // 3
+    private const int MilliSeconds = 3000;
```
コメントと変数名は「何の用途か」がわかるように設定してください。

a-kiyo-g-relations commented on Io.cs:
```diff
@@ -0,0 +1,86 @@
+            Console.WriteLine("あなたの番です");
+            Console.WriteLine("Enterキーでサイコロを投げる");
+            Console.ReadKey();
+        }
```
プレイヤー用とCPU用でメソッドごと分けた方が良いです。
このメソッドに関しては共通化するメリットが特にありません。

a-kiyo-g-relations commented on Io.cs:
```diff
@@ -0,0 +1,86 @@
+    // プレイヤー、役を判断しを出力する
+    public void IsPlayerHand(Player player, Constants.HandRole number, int hand)
+    {
+        if (player.IsPlayer)
```
ここは"あなた"や"相手"の部分も変数にすれば、スマートに分岐できるかと思います。
```c#
string player = "あなた" // 条件によって"相手"に変える

Console.WriteLine($"[{player}]の手は[{hand}]のゾロ目です");
```

a-kiyo-g-relations commented on GameFlow.cs:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    // 役と数字で勝敗を判定する    
+    public void DetermineWinOrLose(Constants.HandRole playerRole, int playerHand,
```
このメソッドですが、以下のように分割してください。
現状だとIOクラス外に表示処理が含まれています。

1. 引数としてplayerとcpuを渡し、勝敗判定を返すメソッド
2. 勝敗の結果を表示するIOクラスのメソッド

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -12,9 +12,13 @@ public class Player
+
+    public bool IsPlayer { get; set; }
+
+    public Player(bool isPlayer)
```
IsPlayerはPlayerクラスに持たせる必要がないかと思います。
基本的には変数の違いで判別できるはずですし、どうしても必要な場合は呼び出し時にboolean型で渡せば良いかと思います。

a-kiyo-g-relations commented on Io.cs:
```diff
@@ -0,0 +1,86 @@
+﻿
+using System;
+
+public class Io
```
今まで普通にIoと呼んでしまっていたのが悪いのですが。。。
「Io」はシステム系のライブラリなどで使われやすい名前なのと、このクラスには業務的な要素も含むためDispなどの方が良いかと思います。

a-kiyo-g-relations replied:
OutputWordsに移植したのであれば、こちらは削除しておいてください。

a-kiyo-g-relations commented on Io.cs:
```diff
@@ -0,0 +1,86 @@
+public class Io
+{
+    // タイトルとゲーム開始を表示する
+    public void GameTitle()
```
このクラスの各メソッドに関しては、staticで良いかと思います。
インスタンスを作成する必要がありません。

a-kiyo-g-relations commented on Io.cs:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    // プレイヤー、役を判断しを出力する
+    public void IsPlayerHand(Player player, Constants.HandRole number, int hand)
```
メソッド名の「Is」はここでは不適切です。

a-kiyo-g-relations replied:
引数を3つ渡していますが、numberは`player.GetHandRoll()`で、handも`player.GetHandValue()`で取得出来るので、個別に渡す必要がありません。

また、呼び出し元ではplayerとcpuのどちらを引数に渡しているか判断出来るため、IsPlayerはPlayerクラスの中に持たせるのではなく、引数でtrue/falseを直接渡せば良いかと思います。

t-nagatsuka-gr committed:
#100 ゲームフロークラスの作成

a-kiyo-g-relations commented on GameFlow.cs:
```diff
@@ -0,0 +1,100 @@
+public class GameFlow
+{
+    // cpuがサイコロを振るまでの待ち時間に必要な値
+    private const int ThreeMinuteValue = 3000;
```
変数名も「3秒間」ではなく「cpuがサイコロを振るまでの待ち時間」を指す名前にしてください。

ちなみにminuteは分です。

a-kiyo-g-relations replied:
こういった時間に関する定義を作る場合、コメントで単位を書いておくと良いです。
```c#
// cpuがサイコロを振るまでの待ち時間(ミリ秒)
private const int ThreeMinuteValue = 3000;
``` 
実業務でもこんな感じで使ったりします。
```c#
// 有効期限(秒)
const int ExpireTime = 24 * 60 * 60;
``` 
この例は有効期限として「1日」を秒換算したものを設定しています。
こうしておけば、後から「有効期限を30分に変えたいけどこの数字って秒単位？分単位？ミリ秒単位？」みたいな事になりません。

a-kiyo-g-relations commented on GameFlow.cs:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    // 役と数字で勝敗を判定する    
+    public void DetermineWinOrLose()
```
このメソッドは
・プレイヤー勝利
・プレイヤー敗北
・引き分け
といういずれかの値を返すメソッドにして、呼び出し元でその戻り値を使って処理を制御した方が良いです。

現在の実装だと判定とその後の処理が一緒になっているので、
例えば「勝ち/引き分けの場合はもう1戦、負けた場合はゲームオーバー」みたいな後続処理の拡張をする際に非常に難しくなってしまいます。

a-kiyo-g-relations replied:
条件式ですが、同じ値の比較は入れ子にするよりelse ifを使った方がわかりやすいかと思います。
```
if (プレイヤーの手役の強さ == CPUの手役の強さ)
{
    if (プレイヤーの出目の強さ > CPUの出目の強さ)
    {
        プレイヤーの勝ち
    }
    else if (プレイヤーの出目の強さ < CPUの出目の強さ)
    {
        プレイヤーの負け
    }
    else
    {
        引き分け
    }
}
```
ifやforなど、入れ子の階層は基本的に深くならない方が良いです。

a-kiyo-g-relations replied:
↑の2つを組み合わせる事でより良い書き方もできますが、まずは上記の修正をしてから説明します。

a-kiyo-g-relations commented on GameFlow.cs:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    // cpuがサイコロを振るまでの待ち時間
+    public void CpuWaitingTime()
```
CpuWaitingTimeだと待ち時間そのものを指しているように見えます。
メソッドは基本的に動詞始まりで定義しましょう。

a-kiyo-g-relations replied:
他メソッド類に関しても「何をする」という部分を動詞形で先頭に付けましょう。
OutputWords内のメソッドなどはほぼ全て「show」や「output」などになりそうです。

a-kiyo-g-relations commented on OutputWords.cs:
```diff
@@ -0,0 +1,106 @@
+﻿
+using System;
+
+public class OutputWords
```
クラス名は単数系にしておきましょう。
今回は関係ありませんが、配列形式でインスタンスを作る時に困ります。
```c#
// 単数か配列か変数名でわからない
OutputWords outputWords;
OutputWords[] outputWords;
```

a-kiyo-g-relations commented on OutputWords.cs:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    // プレイヤー、役を判断しを手の結果を返す
+    public static void ChangePlayerHand(Player player, Player cpu, bool isPlayer)
```
playerとcpuの情報を同時に出力する事はないので、Player型の引数は1つで良いです。
プレイヤーの変数を渡す場合は`ChangePlayerHand(player, true)`のように、CPUの変数を渡す場合は`ChangePlayerHand(cpu, false)`のようにすれば良いです。

a-kiyo-g-relations replied:
更新や変更をする処理はないので、メソッド名の「Change」は不適切です。
コメントの「結果を返す」という部分も間違っています。

a-kiyo-g-relations commented on OutputWords.cs:
```diff
@@ -0,0 +1,106 @@
+    public static void ChangePlayerHand(Player player, Player cpu, bool isPlayer)
+    {
+        string playerName = "あなた";
+        string cpuName = "相手";
```
playerNameの中身をisPlayerの値によって書き換えれば、if文は不要になります。
```
if (isPlayer)
{
    playerName = "あなた"
}
else
{
    playerName = "相手"
}
Console.WriteLine($"{playerName}の手は[{player.GetHandValue()}]のゾロ目です");
``` 
こうすれば「〇〇の手は××です」というメッセージ部分が1箇所で済むので、後から修正する際にやりやすくなります。

a-kiyo-g-relations commented on GameFlow.cs:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    public void Flow()
+    {
```
メソッドの中が長くなっているので、いくつかの処理をサブのメソッドにまとめて分割しておくと良いです。

例えば、こんな分割方法が考えられます。
・タイトル(ゲーム開始前)部分
・プレイヤーの手番
・CPUの手番
・ゲームの結果処理

a-kiyo-g-relations commented on GameFlow.cs:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    // 役と数字で勝敗を判定する    
+    public void DetermineWinOrLose()
```
外部から呼び出すのはFlowメソッドだけなので、このメソッドはprivateで良いです。

private指定には「外部から意図しない呼ばれ方をしない」という利点がありますが、
チーム開発などでは他の開発者に対して「これは内部でしか使わないメソッドです」と明確にできるメリットもあります。

a-kiyo-g-relations commented on OutputWords.cs:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    // プレイヤーの順番であることを表示する
+    public static void PlayerDiceShake(Player player)
```
引数は不要です。

a-kiyo-g-relations commented on OutputWords.cs:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    // cpuの番であることを表示する
+    public static void CpuDiceShake(Player cpu)
```
こちらも引数は不要です。

t-nagatsuka-gr committed:
#100 ゲームフロークラスの作成
第一段階
修正
・OutputWordsクラスのクラス名をOutputWordに変更
・それに伴うGameFlowクラスで記述されているOutputWordの変更

t-nagatsuka-gr committed:
#100 ゲームフロークラスの作成

t-nagatsuka-gr committed:
#100 ゲームフロークラスの作成
・メソッド、メンバ変数の名前修正
(ReturnPlayerHandCompare()は次段階で処理分けをします)

t-nagatsuka-gr committed:
#100 ゲームフロークラスの作成

t-nagatsuka-gr committed:
#100 ゲームフロークラスの作成

a-kiyo-g-relations commented on GameFlow.cs:
```diff
@@ -0,0 +1,88 @@
+        cpu = new Player();
+    }
+
+    public void Flow()
```
このメソッドも動詞始まりにしておきましょう。

a-kiyo-g-relations commented on OutputWord.cs:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    // 役を判断して返す
+    public static void ReturnHandCompare(Player player, bool isPlayer)
```
メソッド名を再確認してください。

a-kiyo-g-relations commented on OutputWord.cs:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    // 役を比較し勝敗を決める    
+    public static Constants.ResultRoll CompareHand(Player player, Player cpu)
```
これは勝敗判定ロジックなので、OutputWord(出力クラス)の中に配置するのは微妙な気がします。

t-nagatsuka-gr committed:
#100 ゲームフロークラスの作成

t-nagatsuka-gr committed:
#100 ゲームフロークラスの作成
